src/Plugin/Commerce/PaymentGateway/Heartland.php

```diff
@@ -0,0 +1,391 @@
+        // If not specified, refund the entire amount.
+        $amount = $amount ?: $payment->getAmount();
+        $this->assertRefundAmount($payment, $amount);
+
+        // Perform the refund request here, throw an exception if it fails.
+        // See \Drupal\commerce_payment\Exception for the available exceptions.
+
+        // Set some variables to use for our transaction
+        $remote_id = $payment->getRemoteId();
+        $number = $amount->getNumber();
+        $currency = $amount->getCurrencyCode();
+
+        // Heartland SDK - Refund Transaction
+        // https://developer.heartlandpaymentsystems.com/Documentation/credit-card-payments/#refund-a-transaction
+        try {
+            $response = Transaction::fromId($remote_id, PaymentMethodType::CREDIT)
+                ->refund($number)
+                ->withCurrency($currency)
+                ->execute();
+  
+            // In case more than one refund has occurred, Commerce plugin sets amounts and states
+            $old_refunded_amount = $payment->getRefundedAmount();
+            $new_refunded_amount = $old_refunded_amount->add($amount);
+            if ($new_refunded_amount->lessThan($payment->getAmount())) {
+                $payment->setState('partially_refunded');
+            } else {
+                $payment->setState('refunded');
+            }
+    
+            $payment->setRefundedAmount($new_refunded_amount);
+            $payment->save();
+        } catch (GatewayException $e) {
+            throw new PaymentGatewayException($e->getMessage());
+        }
+    }
+  
+    /**
+     * {@inheritdoc}
+     */
+    public function createPaymentMethod(PaymentMethodInterface $payment_method, array $payment_details)
+    {
+        $required_keys = [
+            // The expected keys are payment gateway specific and usually match
+            // the PaymentMethodAddForm form elements. They are expected to be valid.
+            'token_value',
+        ];
+        foreach ($required_keys as $required_key) {
+            if (empty($payment_details[$required_key])) {
+                throw new \InvalidArgumentException(sprintf('$payment_details must contain the %s key.', $required_key));
+            }
+        }      
+
+        // Keeps card data from being stored during processing.
+        // Will need to update this for iteration 2, multi-use tokenization.
+        $payment_method->setReusable(false);
+
+        // Set values from form
+        $payment_method->card_type = $payment_details['card_type'];
+        $payment_method->card_number = $payment_details['last_four'];
+        $payment_method->card_exp_month = trim($payment_details['exp_month']);
+        $payment_method->card_exp_year = $payment_details['exp_year'];
+        $expires = CreditCard::calculateExpirationTimestamp(trim($payment_details['exp_month']), $payment_details['exp_year']);
+        $payment_method->setExpiresTime($expires);
```

Comment:
Is this expiry time meant for the token or card?

Reply:
Card.